Query block: client-side pagination

luisherranz opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Built on top of [#53733](https://github.com/WordPress/gutenberg/pull/53733). Supersedes [#44034](https://github.com/WordPress/gutenberg/pull/44034).

This PR adds an option to the Query block to enhance the pagination with client-side navigation.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

To improve the user experience.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

The pagination links are intercepted using a directive (`data-wp-on--click`) which prevents the regular server-side navigation and uses the Interactivity API's `navigate` function instead. It also prefetches the next/previous pages on mouseenter.

## Tasks

There are still many tasks that need to be done before this can be considered ready. I'll probably add more as we progress and find out what needs to be done.

- [x] Only add `supports.interactivity` and enqueue the `view.js` file if the "Enhanced Pagination" option is enabled.
- [x] Add accessibility.
- [x] Scroll to the top of the Query block after navigation.
- [x] Add loading status when loading the next page takes more than 300ms.
- [x] Don't enqueue styles when enhanced pagination is not activated.

### Tasks left for subsequent PRs

- Display a popup when a user tries to enable the enhanced pagination and the Pagination block is not included inside the Query block. The user should be able to choose between canceling the action, or adding the Pagination block (added at the end) using buttons in the popup.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

You can use https://playground.wordpress.net/gutenberg.html to test this PR.

- Add or use a template with the Query block.
- Under the "User Experience" sidebar panel enable the "Enhanced Pagination" setting
- Go to the page/post.
- Use the pagination links.
- The page should not refresh, and the content of the new page should appear on the screen.

## Screenshots or screencast <!-- if applicable -->


https://github.com/WordPress/gutenberg/assets/3305402/e64e7113-62eb-41e2-a67f-e38a0a163534

luisherranz committed:
Add router with region-based client-side navigation

luisherranz committed:
Add changelog

luisherranz committed:
Initial version working

DAreRodz commented on packages/block-library/src/query/edit/inspector-controls/index.js:
```diff
@@ -123,6 +124,18 @@ export default function QueryInspectorControls( props ) {
 		isControlAllowed( allowedControls, 'parents' ) &&
 		isPostTypeHierarchical;
 
+	const enhancedPaginationNotice = __(
+		'Enhanced Pagination might cause interactive blocks within the Post Template to stop working. Disable it if you experience any issues.'
+	);
+
+	const isFirstRender = useRef( true ); // Don't speak on first render.
+	useEffect( () => {
+		if ( ! isFirstRender.current && enhancedPagination ) {
+			speak( enhancedPaginationNotice );
+		}
+		isFirstRender.current = false;
+	}, [ enhancedPagination, enhancedPaginationNotice ] );
```
I need help understanding this part. 😅 This is supposed to prevent announcing the risks associated with the Enhanced Pagination if it's already enabled during the first render, am I right?

Shouldn't we follow the same logic for the `Notice` component rendered below?

If that's the case, maybe we could just simplify the code so we turn `isFirstRender` into regular state and use it along with `enhancedPagination` to conditionally render the `Notice` component, passing the `enhancedPaginationNotice` value as the `spokenMessage`. 🤔

luisherranz replied:
> This is supposed to prevent announcing the risks associated with the Enhanced Pagination if it's already enabled during the first render, am I right?

Yes, that's it 🙂

> Shouldn't we follow the same logic for the `Notice` component rendered below?

No, the notice should appear on the screen, but it should not be announced if the setting is already enabled.

It's basically the same approach that we used here:

- https://github.com/WordPress/gutenberg/pull/52251

carlomanf commented:
This looks good. It would be good to see https://github.com/WordPress/gutenberg/issues/29484 revisited next for intra-post pagination, it was never properly done to begin with.

luisherranz committed:
Merge branch 'trunk' into add/interactivity-api-regions-based-client-side-navigation

luisherranz committed:
Only add supports.interactivity and enqueue the view.js file if the "Enhanced Pagination" option is enabled

luisherranz committed:
Interactivity API: merge new server-side rendered context on client-side navigation (#53853)

* Add failing test

* Fix the test

* Add changelog

* Fix lint error

luisherranz committed:
Fix changelog placement

luisherranz committed:
Interactivity API: Support for the `data-wp-key` directive (#53844)

* Add failing test

* Fix test using key

* Replace key with data-wp-key

* Refactor test a bit

* Add changelog

* Add docs

* Remove unnecessary paragraph

* Fix lint error

luisherranz committed:
Merge branch 'add/interactivity-api-regions-based-client-side-navigation' into try/client-side-pagination-of-query-block

luisherranz committed:
Interactivity API: Fix non stable context reference on client side navigation (#53876)

* Add failing test

* Fix the test

luisherranz committed:
Merge branch 'add/interactivity-api-regions-based-client-side-navigation' into try/client-side-pagination-of-query-block

luisherranz committed:
Add basic accessibility (announcement + focus)

luisherranz committed:
Add basic loading animation

Co-authored-by: David Arenas <darerodz@users.noreply.github.com>

luisherranz committed:
Only enqueue styles if they are needed

luisherranz committed:
Add missing CSS props on aria-live element

Apparently, some screen readers don't read elements that don't have width and height.

luisherranz commented:
This is still the most basic version of this functionality, but it should be ready for the first reviews.

@darerodz: would you mind taking a look at the code? Thanks.

---

@alexstine, @joedolson, @jeryj: would you mind taking a look at the accessibility?

This is how it currently works:

- When the navigation starts, it waits 300ms.
- After 300ms, if the navigation hasn't finished, it announces that the page is loading: `Loading page, please wait.`
- When the navigation ends, it announces that the new page is ready (`The new page has been loaded. You can continue navigating.`) and it positions the focus on the first post of the new page.

The 300ms delay is used to avoid announcing the `Loading page, please wait.` message when the connection is very fast or the page is cached because it has been visited before.

There's no need for error control because if something fails or a timeout is reached, we will immediately fall back to server-side navigation. _That is coming on a separate PR._

Please, let me know if something is missing or if I should change something. Thanks a lot 🙏🙂

luisherranz commented on packages/block-library/src/query/index.php:
```diff
@@ -5,12 +5,105 @@
+			);
+			$content = $p->get_updated_html();
+
+			// Make the block as interactive.
```
```suggestion
			// Mark the block as interactive.
```

luisherranz commented on packages/block-library/src/query/index.php:
```diff
@@ -5,12 +5,105 @@
+			$p->set_attribute( 'data-wp-navigation-id', 'query-' . $attributes['queryId'] );
+			$p->set_attribute(
+				'data-wp-context',
+				json_encode( array( 'core' => array( 'query' => (object) array() ) ) )
```
```suggestion
				wp_json_encode( array( 'core' => array( 'query' => (object) array() ) ) )
```

luisherranz committed:
Fix two typos

pablohoneyhoney commented:
Very nice. Would we also add some swift fading to make the transition a bit more _kansei_? even if the user prefers to deactivate it. And perhaps UX label is a bit misleading? UX is everything.

jeryj commented on packages/block-library/src/query/index.php:
```diff
@@ -5,12 +5,105 @@
+			$content           = substr_replace(
+				$content,
+				'<div
+					style="position:absolute;clip:rect(0,0,0,0);width:1px;height:1px;padding:0;margin:-1px;overflow:hidden;border:0;"
```
Is there a CSS class we could use here?

DAreRodz replied:
Fixed in https://github.com/WordPress/gutenberg/pull/53812/commits/6c680a61261af5ad6fbc8378cb67305557c02be3.

jeryj commented on packages/block-library/src/query/index.php:
```diff
@@ -5,12 +5,105 @@
+						'core' => array(
+							'query' => array(
+								'loadingText' => __( 'Loading page, please wait.' ),
+								'loadedText'  => __( 'The new page has been loaded. You can continue navigating.' ),
```
"Page Loaded" is probably enough here. Screen readers announce a lot. The fewer words we can use to communicate the same thing, the better.

Also, I believe some kind of announcement of the dynamically loaded content is helpful. Maybe wrapping the posts container in an aria-live? Not positive though.

DAreRodz replied:
> "Page Loaded" is probably enough here.

Fixed in https://github.com/WordPress/gutenberg/pull/53812/commits/fce30323e66d2ebd2b0c25c298e10325840eea47.

> Also, I believe some kind of announcement of the dynamically loaded content is helpful.

I will do this in a follow-up PR. 👍

DAreRodz replied:
Hi, @jeryj. I was about to open a PR to address these changes. 🙂

> Maybe wrapping the posts container in an aria-live?

I just tried so, but the screen reader announces lots of changes (the whole list, basically). Is there a way to summarize the changes somehow? 😅

jeryj replied:
Yeah, I was probably wrong to have it be wrapped in `aria-live`. I know you can press `ctrl` to stop the announcements though if you don't want to hear all of them. Maybe it could announce the post titles of all the fetched items? 

@alexstine, do you have guidance for what the preferred pattern is? I've looked around and couldn't find much.

alexstine replied:
I think just the loading announcement is fine followed by focus on the first item.

DAreRodz replied:
Thanks for your feedback, @jeryj and @alexstine. 🙏

> I think just the loading announcement is fine followed by focus on the first item.

This is the current implementation; no further action would be needed, so I'll set this conversation as resolved. Feel free to reopen it if you think the accessibility could be handled differently. 🙂

jeryj commented on packages/block-library/src/query/view.js:
```diff
@@ -0,0 +1,82 @@
+
+						context.core.query.animation = 'finish';
+
+						// Focus the first anchor of the Query block.
+						document
+							.querySelector(
+								`[data-wp-navigation-id=${ id }] a[href]`
+							)
+							?.focus();
```
We should manage the focus to return to the clicked button so we're not stealing focus to a new place. The `current-page` element can remain an `<a>` element so it can receive focus and simplify the code. I'm not 100% positive of this interaction though. I think it would be better than always returning to the first item though.

DAreRodz replied:
We will address this in a follow-up PR. 👍

DAreRodz replied:
Hi, @jeryj. I was about to open a PR to address these changes. 🙂 

You propose to keep a link to the current page with `aria-current=page`, and maintain focus there, am I right? There is a case I don't know how to manage: when the Pagination Number block is not present. What would be the best approach in this case?

jeryj replied:
I'd say to return focus to whatever was triggered, unless it is now disabled, such as if you click a Next button and there's no more page. In that case, probably put it on the closest sensible button, such as the Previous button? I'm not positive here either.

alexstine replied:
This is a tricky one because focus probably needs to go back to the newly loaded results. One could make an argument that this makes the experience worse because you could not keep hitting next but you also can't do this when the page reloads either.

jeryj replied:
Moving focus to the results after load likely simplifies the code and makes the focus experience more consistent, so sounds good to me! Thanks!

DAreRodz replied:
Again, thanks for your feedback, @jeryj and @alexstine. 😊

This is also the current implementation, so no more changes are needed. Setting this conversation as resolved.

DAreRodz committed:
Merge branch 'trunk' into try/client-side-pagination-of-query-block

DAreRodz committed:
Update fixture

DAreRodz committed:
Simplify loaded text

DAreRodz committed:
Move navigation announce style to a class

DAreRodz committed:
Merge branch 'trunk' into try/client-side-pagination-of-query-block

DAreRodz reviewed (approved):
Code LGTM! It also behaves as expected. Approved! 🚀 

PS: I'm aware there are still some issues to address (thanks, @pablohoneyhoney and @jeryj for the feedback 😊 ). We will focus on that in subsequent PRs.

alexstine commented:
As I noted in my previous comment on #53737, what is the plan for addressing these follow-ups? Memory is not good enough. This is how things get forgotten and go unaddressed.

I am still trying to catch up after WCUS, haven't had time to review much.

jeryj commented:
Commenting to support @alexstine here. Merging when it doesn't meet accessibility standards is merging in a broken state. It can take some time to make the mental shift, but please consider screen reader, keyboard, and other ATs as baselines rather than follow-ups. For example, I doubt this PR would have been merged if the pagination buttons didn't work via a mouse click. In the same way, it shouldn't have been merged without proper focus management.

I'm not trying to be hard on y'all. I'm very glad you are all considering accessibility as important. I do want to get on the same page as to what is considered ready for merge or not though so we can have the same expectations.

DAreRodz commented:
First of all, thanks a lot for your feedback and for sharing your concerns. I totally understand it, and I agree we should work on solving them. I had assumed that the noted accessibility issues were not as blocking and could be fixed in a subsequent PR before the coming release. Sorry for the rush to merge the PR; I should have spent more time on accessibility. 🙏

I already included them in the [Tracking Issue](https://github.com/WordPress/gutenberg/issues/53740), and I plan to work on them ASAP. I believe we can work on a PR to fix them during the RC to ensure it works fine before this Gutenberg release. 😊

jameskoster commented:
Was there any discussion about putting this option in the Settings panel of the Pagination block? 

Not all queries include pagination, so finding the setting on the Query block itself might feel a bit strange in those circumstances. "User experience" is an ambiguous panel title, and generally we try to avoid ad hoc panels for narrow use cases.

DAreRodz commented:
> Was there any discussion about putting this option in the Settings panel of the Pagination block?

Yep, there's already a PR opened to solve that:

- https://github.com/WordPress/gutenberg/pull/54198

> Not all queries include pagination, so finding the setting on the Query block itself might feel a bit strange in those circumstances.

Uh, you're right. I think @luisherranz also noticed the same, as he suggested adding a pop-up to warn about activating the "enhanced pagination" when there's no Pagination block. It's mentioned under **Tasks left for subsequent PRs**, in the PR's description.

@luisherranz, I guess we should move that task to the [Tracking Issue](https://github.com/WordPress/gutenberg/issues/53740), right?

> "User experience" is an ambiguous panel title, and generally we try to avoid ad hoc panels for narrow use cases.

Agree, and @pablohoneyhoney mentioned something similar in https://github.com/WordPress/gutenberg/pull/53812#issuecomment-1693305042. It is also addressed in https://github.com/WordPress/gutenberg/pull/54198.

luisherranz commented:
> Not all queries include pagination, so finding the setting on the Query block itself might feel a bit strange in those circumstances.

We put it in the Query block because when this is activated, we need to add directives to the Query, Post Template and Pagination blocks.

But if there's a way to set the block context in the Pagination block and read it in the Query and Post Template blocks, we'll happily move the setting there 🙂

luisherranz commented:
> @luisherranz, I guess we should move that task to the Tracking Issue, right?

True. I've just added it!
